### ISSUE TITLE
Scylla crashes with `std::bad_alloc` error

### ISSUE THREAD

**Issue by PleasingFungus** [scylladb/scylla#6110]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [X] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 3.1.0-0.20191012.9c3cdded9
Cluster size: 14 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu

During routine operations (no human intervention, no heavy load), Scylla crashed with this error:

```
2020/03/31 07:56:47 terminate called after throwing an instance of 'std::bad_alloc'
2020/03/31 07:56:47   what():  std::bad_alloc
2020/03/31 07:56:47 Aborting on shard 23.
```

Full logs for the day: [scylla-current.log](https://github.com/scylladb/scylla/files/4411488/scylla-current.log).

Running `nodetool status` on the crashed node:

```
$ nodetool status
/usr/bin/filter_cassandra_attributes.py:10: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  attributes.update(load(open(sys.argv[i], 'r')))
nodetool: Unable to connect to Scylla API server: java.net.ConnectException: Connection refused (Connection refused)
See 'nodetool help' or 'nodetool help <command>'.
```

`nodetool status` on another node:
```
$ nodetool status
/usr/bin/filter_cassandra_attributes.py:10: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  attributes.update(load(open(sys.argv[i], 'r')))
Datacenter: us-east
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address        Load       Tokens       Owns    Host ID                               Rack
UN  172.20.4.246   6.53 TB    256          ?       9528dd3c-9b6d-4f68-b93a-8fcb133219ad  1d
UN  172.20.62.86   6.14 TB    256          ?       9706e8c1-5ea5-4f09-8478-47fe1bfc898d  1d
UN  172.20.64.49   6.12 TB    256          ?       50def455-c660-434c-8a72-a78ffd6de30d  1d
UN  172.20.96.51   6.35 TB    256          ?       ffaf433a-b837-47b7-b103-b41c9e574fcf  1d
UN  172.20.87.188  5.57 TB    256          ?       2ea5ec11-f6dd-4eaf-ac8e-42253cea2827  1d
UN  172.20.86.29   6.02 TB    256          ?       4c4b1f54-86d4-4286-8a47-f1fe27fbb653  1d
UN  172.20.37.184  6.49 TB    256          ?       fb0f4794-d9a5-4904-b9df-0ebfb51c43b0  1d
UN  172.20.48.219  6.56 TB    256          ?       4430bd1a-c386-412f-a51d-9e4880bd8280  1d
UN  172.20.41.198  6.42 TB    256          ?       a1a34cd8-99c5-4d26-b759-a4f32145a2f1  1d
DN  172.20.62.204  6.67 TB    256          ?       551b472a-969a-4160-9a27-051859b23635  1d
UN  172.20.31.77   6.43 TB    256          ?       159a2dcd-b872-4bd9-9e95-421f8a5b9047  1d
UN  172.20.97.206  5.97 TB    256          ?       2a1df79e-82ec-49bb-b5a9-57ed21f5ad84  1d
UN  172.20.94.40   6.78 TB    256          ?       b1b7341d-e0bf-435e-a966-a547d97ad80a  1d
UN  172.20.33.203  6.57 TB    256          ?       d7b32443-692f-47df-a427-ad851f55d41e  1d
```

```
$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev             94G     0   94G   0% /dev
tmpfs            19G  844K   19G   1% /run
/dev/nvme2n1p1  7.7G  2.7G  5.0G  36% /
tmpfs            94G     0   94G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs            94G     0   94G   0% /sys/fs/cgroup
/dev/md0         14T  6.9T  6.1T  54% /mnt
/dev/loop3       18M   18M     0 100% /snap/amazon-ssm-agent/1480
/dev/loop0       92M   92M     0 100% /snap/core/8689
/dev/loop2       18M   18M     0 100% /snap/amazon-ssm-agent/1566
/dev/loop1       94M   94M     0 100% /snap/core/8935
tmpfs            19G     0   19G   0% /run/user/1000
```

Graphs leading up to the crash:

![Screen Shot 2020-03-31 at 1 43 17 PM](https://user-images.githubusercontent.com/1192475/78073355-99ebc000-7355-11ea-9f81-f6a0dc1983a0.png)
![Screen Shot 2020-03-31 at 1 43 12 PM](https://user-images.githubusercontent.com/1192475/78073357-9b1ced00-7355-11ea-90a4-71a091c764cc.png)
![Screen Shot 2020-03-31 at 1 43 09 PM](https://user-images.githubusercontent.com/1192475/78073360-9c4e1a00-7355-11ea-9cc9-3904ca5fc378.png)
![Screen Shot 2020-03-31 at 1 43 03 PM](https://user-images.githubusercontent.com/1192475/78073361-9ce6b080-7355-11ea-949a-91feb20eedb0.png)
![Screen Shot 2020-03-31 at 1 42 53 PM](https://user-images.githubusercontent.com/1192475/78073363-9d7f4700-7355-11ea-9c8b-155156c9d7ea.png)
![Screen Shot 2020-03-31 at 1 42 48 PM](https://user-images.githubusercontent.com/1192475/78073364-9d7f4700-7355-11ea-98c7-8e3730f9b335.png)

I'm happy to provide more details.

**Comment by PleasingFungus** [comment]
Per request, config files from `/etc/scylla.d/` and `scylla.yaml`:
[scyllaconf.zip](https://github.com/scylladb/scylla/files/4411549/scyllaconf.zip)

**Comment by PleasingFungus** [comment]
Per requests, per-shard and per-instance graphs.

![Screen Shot 2020-03-31 at 1 53 23 PM](https://user-images.githubusercontent.com/1192475/78074419-922d1b00-7357-11ea-8f35-12de76506266.png)
![Screen Shot 2020-03-31 at 1 53 29 PM](https://user-images.githubusercontent.com/1192475/78074424-935e4800-7357-11ea-88c7-2173e8f6c089.png)
![Screen Shot 2020-03-31 at 1 53 36 PM](https://user-images.githubusercontent.com/1192475/78074426-935e4800-7357-11ea-8220-bb9e241b02dd.png)
![Screen Shot 2020-03-31 at 1 53 41 PM](https://user-images.githubusercontent.com/1192475/78074428-93f6de80-7357-11ea-80ea-9729eba71505.png)
![Screen Shot 2020-03-31 at 1 53 45 PM](https://user-images.githubusercontent.com/1192475/78074431-948f7500-7357-11ea-8222-e48bcef6845a.png)
![Screen Shot 2020-03-31 at 1 54 01 PM](https://user-images.githubusercontent.com/1192475/78074432-95280b80-7357-11ea-84b5-8d4956652aad.png)
![Screen Shot 2020-03-31 at 1 57 04 PM](https://user-images.githubusercontent.com/1192475/78074434-95c0a200-7357-11ea-8131-79ec984303d4.png)
![Screen Shot 2020-03-31 at 1 57 08 PM](https://user-images.githubusercontent.com/1192475/78074435-96593880-7357-11ea-85ae-0d6ce9c832ae.png)
![Screen Shot 2020-03-31 at 1 57 11 PM](https://user-images.githubusercontent.com/1192475/78074437-96593880-7357-11ea-888d-216c59e1fd3d.png)
![Screen Shot 2020-03-31 at 1 57 15 PM](https://user-images.githubusercontent.com/1192475/78074445-99542900-7357-11ea-96a7-b38dc1649d3b.png)
![Screen Shot 2020-03-31 at 1 57 21 PM](https://user-images.githubusercontent.com/1192475/78074448-99ecbf80-7357-11ea-9b18-b77f4b34341c.png)

**Comment by dorlaor** [comment]
It's an out of memory. Is there a reason why - do you have large wide
partitions
or extremely large collections/blobs? I assume it's not reproducible that
easily.
Getting the backtrace with debug symbols will help.

**Comment by raphaelsc** [comment]
@PleasingFungus could you please share output of ` find /mnt/scylla/data/ -type f -printf '%f %s\n'` and also all schemas in tribute keyspace?

**Comment by denesb** [comment]
@PleasingFungus can you please upload the coredump as well? See instructions here https://docs.scylladb.com/troubleshooting/report_scylla_problem/#core-dump.

**Comment by PleasingFungus** [comment]
@dorlaor How would I get the backtrace with debug symbols?

@raphaelsc [Output of find](https://github.com/scylladb/scylla/files/4416584/out.txt).

(Note that the server is back up and running now, so this doesn't perfectly reflect the state the server was in when it crashed. Sorry!)

Schemas:

```
CREATE KEYSPACE tribute WITH replication = {'class': 'NetworkTopologyStrategy', 'us-east': '3'}  AND durable_writes = true;

CREATE TABLE tribute.impressions7d (
    k blob PRIMARY KEY,
    txid bigint,
    v blob,
    version int
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '240', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.2'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 604800
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE tribute.impressions1d (
    k blob PRIMARY KEY,
    txid bigint,
    v blob,
    version int
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '30', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.2'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 86400
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

@denesb I can't find a core dump, unfortunately. `/var/lib/scylla/coredump/` exists, but the directory is empty. `/etc/systemd/coredump.conf.d/custom.conf` also doesn't exist - does that mean that Scylla won't generate core dumps?

Thanks all for the quick responses!

**Comment by dorlaor** [comment]
https://docs.scylladb.com/kb/decode-stack-trace/

**Comment by PleasingFungus** [comment]
Following the instructions on the page, I get:

```
$ sudo apt-get install scylla-server-dbg
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package scylla-server-dbg
```

Is there a repo I need to add? Can I download the debian package directly from somewhere?

**Comment by dorlaor** [comment]
@yaronkaikov do we have debug packages for 3.1 ?
I hope we do ship them with newer releases.
@PleasingFungus until you get help and regardless, I suggest to upgrade to a newer
release since 3.1 became end of life with the introduction of 3.3

**Comment by PleasingFungus** [comment]
OK, I'll take a look at upgrading.

**Comment by PleasingFungus** [comment]
https://www.scylladb.com/download/open-source/#binary says:

> ScyllaDB maintains the current and previous releases of Scylla Open Source. We currently maintain Scylla Open Source 3.2.x and 3.1.x.

Is that out of date?

**Comment by yaronkaikov** [comment]
> @yaronkaikov do we have debug packages for 3.1 ?
> I hope we do ship them with newer releases.
> @PleasingFungus until you get help and regardless, I suggest to upgrade to a newer
> release since 3.1 became end of life with the introduction of 3.3

@dorlaor did you mean to ask me or Yaron from QA?

**Comment by PleasingFungus** [comment]
Unfortunately, that node failed last night and had to be taken out of the cluster. When I bring the node back up and run that query on it, I don't see anything.

```
 keyspace_name | table_name | sstable_name | partition_size | partition_key | compaction_time
---------------+------------+--------------+----------------+---------------+-----------------

(0 rows)
```

**Comment by raphaelsc** [comment]
@PleasingFungus I'd highly recommend you to consider aligning your partition keys to your time windows. 1) Read will touch only 1 sstable. 2) Bucketing partitions keep their size reasonable, which brigs lots of space and perf advantages.

For the table with 30 min window, you could put some field as part of primary key which will uniquely identify that window. For example, a table with 1 day window, you would put something like day of year as part of primary key.

I think most of your performance issues have to do with large partitions.

**Comment by PleasingFungus** [comment]
OK, I'll try that and report back. Thank you for the advice!

**Comment by PleasingFungus** [comment]
To be clear, are you suggesting something like:

```
CREATE TABLE tribute.impressions1d (
    k blob,
    time_bucket timestamp,
    txid bigint,
    v blob,
    version int,
    PRIMARY KEY (time_bucket, k)
...

CREATE INDEX ON tribute.impressions1d(k);
```

That is, using "time rounded to the compaction window size" (`time_bucket`) as the partition key, and the original key `k` (used for queries) as a clustering column with an index to allow for queries for the form `SELECT txid, v, version FROM tribute.impressions1d WHERE k = foo`?

If so, how would you suggest I test for performance improvements?

**Comment by raphaelsc** [comment]
On Tue, Apr 7, 2020 at 5:57 PM Nicholas Feinberg <notifications@github.com>
wrote:

> To be clear, are you suggesting something like:
>
> CREATE TABLE tribute.impressions1d (
>     k blob,
>     time_bucket timestamp,
>     txid bigint,
>     v blob,
>     version int,
>     PRIMARY KEY (time_bucket, k)
> ...
>
> CREATE INDEX ON tribute.impressions1d(k);
>
> That is, using "time rounded to the compaction window size" (time_bucket)
> as the partition key, and the original key k (used for queries) as a
> clustering column with an index to allow for queries for the form SELECT
> txid, v, version FROM tribute.impressions1d WHERE k = foo?
>
> If so, how would you suggest I test for performance improvements?
>
I'd suggest we look at cache metrics, IOPS, and RPS. By aligning partitions
to time windows, more data can be cached and a miss will result in only 1
sstable read in average, which translates into improved read performance.

**Comment by PleasingFungus** [comment]
OK, thanks!

**Comment by PleasingFungus** [comment]
@raphaelsc I tried this suggestion. The resulting schema was:

```
CREATE KEYSPACE tribute WITH replication = {'class': 'NetworkTopologyStrategy', 'us-east': '3'}  AND durable_writes = true;

CREATE TABLE tribute.impressions7d (
    time_bucket timestamp PRIMARY KEY,
    k blob,
    txid bigint,
    v blob,
    version int
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '240', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.2'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 604800
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
CREATE INDEX impressions7d_k_idx ON tribute.impressions7d (k);

CREATE MATERIALIZED VIEW tribute.impressions7d_k_idx_index AS
    SELECT k, idx_token, time_bucket
    FROM tribute.impressions7d
    WHERE k IS NOT NULL
    PRIMARY KEY (k, idx_token, time_bucket)
    WITH CLUSTERING ORDER BY (idx_token ASC, time_bucket ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE tribute.impressions1d (
    time_bucket timestamp PRIMARY KEY,
    k blob,
    txid bigint,
    v blob,
    version int
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '30', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.2'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 86400
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
CREATE INDEX impressions1d_k_idx ON tribute.impressions1d (k);

CREATE MATERIALIZED VIEW tribute.impressions1d_k_idx_index AS
    SELECT k, idx_token, time_bucket
    FROM tribute.impressions1d
    WHERE k IS NOT NULL
    PRIMARY KEY (k, idx_token, time_bucket)
    WITH CLUSTERING ORDER BY (idx_token ASC, time_bucket ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

This seems to produce performance that is dramatically worse. The great majority of writes are localized onto a single node, causing large numbers of timeouts.

```
$ nodetool status
/usr/bin/filter_cassandra_attributes.py:10: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  attributes.update(load(open(sys.argv[i], 'r')))
Datacenter: us-east
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address         Load       Tokens       Owns    Host ID                               Rack
UN  172.20.51.39    530.36 KB  256          ?       d4ea5f21-4796-414d-8192-68ab31dfadc4  1d
UN  172.20.33.224   4.94 MB    256          ?       4aa37b70-2494-423b-85f6-b84cb2ea7083  1d
UN  172.20.102.242  10.16 MB   256          ?       0c5e0646-0f21-4ab4-b8a7-2dce6d46322f  1d
UN  172.20.48.31    505.52 KB  256          ?       d30b6063-8cdb-4cf6-b4f9-5deee0c91e03  1d
UN  172.20.104.121  5.13 MB    256          ?       4d35b049-bc65-4015-87ba-65c35117fe7c  1d
UN  172.20.37.123   5.74 MB    256          ?       8cf99192-a025-45c4-ae4d-d1f8afa388c7  1d
UN  172.20.33.155   5.48 MB    256          ?       47d83483-f015-460f-bf44-ba49f46c1687  1d
```

![Screen Shot 2020-04-10 at 3 49 54 PM](https://user-images.githubusercontent.com/1192475/79027925-f19ede00-7b42-11ea-8b3e-c39fdb6d413e.png)
![Screen Shot 2020-04-10 at 3 49 42 PM](https://user-images.githubusercontent.com/1192475/79027929-f368a180-7b42-11ea-9636-9eb2ff1d9b2f.png)
![Screen Shot 2020-04-10 at 3 49 29 PM](https://user-images.githubusercontent.com/1192475/79027931-f499ce80-7b42-11ea-9e85-eb6477be63ec.png)

My guess is that, by moving all writes for a given time period into a single SSTable, we also move them onto a single machine, resulting in very poor load balancing within the fleet. Does that seem plausible to you?

Thanks again,
Nicholas

**Comment by slivne** [comment]
@PleasingFungus - the change you did 

CREATE TABLE tribute.impressions7d (
    time_bucket timestamp PRIMARY KEY,
    k blob,
    txid bigint,
    v blob,
    version int
);

doesn't align with 
CREATE TABLE tribute.impressions1d (
    k blob,
    time_bucket timestamp,
    txid bigint,
    v blob,
    version int,
    PRIMARY KEY (time_bucket, k)
);


your change created a single partition that is allways overwritten for a single time_bucket id - was this the intent ?

**Comment by PleasingFungus** [comment]
> your change created a single partition that is allways overwritten for a single time_bucket id - was this the intent ?

I'm not sure what you mean by that, sorry. Can you rephrase?

My intent was to follow Raphael's advice [here](https://github.com/scylladb/scylla/issues/6110#issuecomment-610126548).

**Comment by slivne** [comment]
For this  schema 

```
CREATE TABLE tribute.impressions7d (
time_bucket timestamp PRIMARY KEY,
k blob,
txid bigint,
v blob,
version int
);
```
(there is no clustering_key)

sample inserts

insert into tribute.impressions7d (time_bucket, k) values (11,0x00)
insert into tribute.impressions7d (time_bucket, k) values (11,0x11)

will result in a single partition holding only time_bucket = 11, k= 0x11. 

since there is no clustering_key.

**Comment by PleasingFungus** [comment]
Ah, I think I see. I'll try it again using the clustering key, then - thanks!

**Comment by PleasingFungus** [comment]
OK, I tried using the clustering key. Specifically, this schema:

```
CREATE KEYSPACE tribute WITH replication = {'class': 'NetworkTopologyStrategy', 'us-east': '3'}  AND durable_writes = true;

CREATE TABLE tribute.impressions7d (
    time_bucket timestamp,
    k blob,
    txid bigint,
    v blob,
    version int,
    PRIMARY KEY (time_bucket, k)
) WITH CLUSTERING ORDER BY (k ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '240', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.2'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 604800
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
CREATE INDEX impressions7d_k_idx ON tribute.impressions7d (k);

CREATE MATERIALIZED VIEW tribute.impressions7d_k_idx_index AS
    SELECT k, idx_token, time_bucket
    FROM tribute.impressions7d
    WHERE k IS NOT NULL
    PRIMARY KEY (k, idx_token, time_bucket)
    WITH CLUSTERING ORDER BY (idx_token ASC, time_bucket ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

CREATE TABLE tribute.impressions1d (
    time_bucket timestamp,
    k blob,
    txid bigint,
    v blob,
    version int,
    PRIMARY KEY (time_bucket, k)
) WITH CLUSTERING ORDER BY (k ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '30', 'compaction_window_unit': 'MINUTES', 'tombstone_threshold': '0.2'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 86400
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
CREATE INDEX impressions1d_k_idx ON tribute.impressions1d (k);

CREATE MATERIALIZED VIEW tribute.impressions1d_k_idx_index AS
    SELECT k, idx_token, time_bucket
    FROM tribute.impressions1d
    WHERE k IS NOT NULL
    PRIMARY KEY (k, idx_token, time_bucket)
    WITH CLUSTERING ORDER BY (idx_token ASC, time_bucket ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
``` 

Please let me know if I overlooked anything again.

When using this schema, I encountered extremely high error rates on writes. I've attached a representative log: [scylla-current.log.zip](https://github.com/scylladb/scylla/files/4490040/scylla-current.log.zip)

```
$ nodetool status
/usr/bin/filter_cassandra_attributes.py:10: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  attributes.update(load(open(sys.argv[i], 'r')))
Datacenter: us-east
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address         Load       Tokens       Owns    Host ID                               Rack
UN  172.20.51.39    14.78 GB   256          ?       d4ea5f21-4796-414d-8192-68ab31dfadc4  1d
UN  172.20.33.224   18.8 GB    256          ?       4aa37b70-2494-423b-85f6-b84cb2ea7083  1d
UN  172.20.102.242  11.86 GB   256          ?       0c5e0646-0f21-4ab4-b8a7-2dce6d46322f  1d
UN  172.20.48.31    7.76 GB    256          ?       d30b6063-8cdb-4cf6-b4f9-5deee0c91e03  1d
UN  172.20.104.121  21.73 GB   256          ?       4d35b049-bc65-4015-87ba-65c35117fe7c  1d
UN  172.20.37.123   15.09 GB   256          ?       8cf99192-a025-45c4-ae4d-d1f8afa388c7  1d
UN  172.20.33.155   15.93 GB   256          ?       47d83483-f015-460f-bf44-ba49f46c1687  1d
```

![Screen Shot 2020-04-16 at 3 38 20 PM (3)](https://user-images.githubusercontent.com/1192475/79513642-302c1100-7ff9-11ea-816d-c9a38870f569.png)
![Screen Shot 2020-04-16 at 3 39 02 PM](https://user-images.githubusercontent.com/1192475/79513648-33270180-7ff9-11ea-9456-94962c01cefb.png)
![Screen Shot 2020-04-16 at 3 38 26 PM](https://user-images.githubusercontent.com/1192475/79513654-33bf9800-7ff9-11ea-830f-636bd5aea95e.png)
![Screen Shot 2020-04-16 at 3 38 03 PM](https://user-images.githubusercontent.com/1192475/79513655-33bf9800-7ff9-11ea-9a5d-a2950bd76201.png)
![Screen Shot 2020-04-16 at 3 38 52 PM](https://user-images.githubusercontent.com/1192475/79513656-34582e80-7ff9-11ea-89a8-4ee3c0607e68.png)

The mentions of 'large partitions' and 'sstable read queue overloaded' both seem potentially relevant. Any idea what might be happening here?

**Comment by avikivity** [comment]
The interesting metric is "Foreground Writes Per Instance". Can you show it per shard too (for one instance at a time).

Also, can you upload the metrics database? Hopefully the crash time is still retained there.

**Comment by PleasingFungus** [comment]
> The interesting metric is "Foreground Writes Per Instance". Can you show it per shard too (for one instance at a time). 

Sure! Let me know if this is what you wanted:

<img width="1666" alt="Screen Shot 2020-04-19 at 12 00 53 PM" src="https://user-images.githubusercontent.com/1192475/79697105-a1f09e80-8235-11ea-9862-093085e194c3.png">
<img width="1666" alt="Screen Shot 2020-04-19 at 12 00 35 PM" src="https://user-images.githubusercontent.com/1192475/79697109-a6b55280-8235-11ea-8454-2c1a9f8e801f.png">
<img width="1664" alt="Screen Shot 2020-04-19 at 12 00 27 PM" src="https://user-images.githubusercontent.com/1192475/79697111-a74de900-8235-11ea-89aa-0e37dfe8811c.png">
<img width="1669" alt="Screen Shot 2020-04-19 at 12 00 16 PM" src="https://user-images.githubusercontent.com/1192475/79697114-a7e67f80-8235-11ea-9407-ce33158de3b8.png">
<img width="1664" alt="Screen Shot 2020-04-19 at 12 01 02 PM" src="https://user-images.githubusercontent.com/1192475/79697106-a4eb8f00-8235-11ea-8b57-28ff8a2d2f57.png">
<img width="1671" alt="Screen Shot 2020-04-19 at 12 00 44 PM" src="https://user-images.githubusercontent.com/1192475/79697107-a5842580-8235-11ea-8aee-f33ef7bfd28f.png">
<img width="1662" alt="Screen Shot 2020-04-19 at 12 01 11 PM" src="https://user-images.githubusercontent.com/1192475/79697108-a61cbc00-8235-11ea-96df-d6767299b1f2.png">

> Also, can you upload the metrics database? Hopefully the crash time is still retained there.

Do you mean the Prometheus DB that was storing data for the old instances? That's been gone for a bit, sorry :(

**Comment by avikivity** [comment]
Use "by: shard" instead of "by: instance" in the top-left corner. But I guess this time period doesn't include a crash.

**Comment by PleasingFungus** [comment]
Apologies for the delay here - just got back from vacation! Here's what I believe you're asking for, but please correct me if I misunderstood again.

![Screen Shot 2020-05-01 at 2 51 42 PM](https://user-images.githubusercontent.com/1192475/80844385-5afe9380-8bbb-11ea-86ca-6a79beb9aabc.png)
![Screen Shot 2020-05-01 at 2 51 37 PM](https://user-images.githubusercontent.com/1192475/80844390-5df98400-8bbb-11ea-9a3e-14674744ee34.png)
![Screen Shot 2020-05-01 at 2 51 32 PM](https://user-images.githubusercontent.com/1192475/80844391-5e921a80-8bbb-11ea-9c62-e9ec26c0ae0c.png)
![Screen Shot 2020-05-01 at 2 51 27 PM](https://user-images.githubusercontent.com/1192475/80844394-5f2ab100-8bbb-11ea-9751-48e947ecab74.png)
![Screen Shot 2020-05-01 at 2 51 22 PM](https://user-images.githubusercontent.com/1192475/80844397-605bde00-8bbb-11ea-9ac7-7314ab48a653.png)
![Screen Shot 2020-05-01 at 2 51 16 PM](https://user-images.githubusercontent.com/1192475/80844399-605bde00-8bbb-11ea-812e-12360300c39f.png)
![Screen Shot 2020-05-01 at 2 51 10 PM](https://user-images.githubusercontent.com/1192475/80844404-618d0b00-8bbb-11ea-918c-57776d533ec5.png)

Did you have any hypotheses for what's going wrong, either with the crash described here or the possibly-associated performance issues? If you have any ideas for investigations I could run on my end, I'm all ears.

**Comment by slivne** [comment]
@PleasingFungus is this still happening on latest OSS releases of scylla 4.0 / 4.1 - if so can you please open a new bug and reference this issue and we will work with you to fix it.
